ecosystem/indexer-grpc/indexer-grpc-data-service/src/service.rs

```diff
@@ -425,6 +198,355 @@ impl RawData for RawDataServerWrapper {
     }
 }
 
+enum DataFetchSubThreadResult {
+    // This is the first/last txn version returned by this
+    Success(Vec<Transaction>),
+    NoResults,
+}
+
+async fn get_data_with_threads(
+    start_version: u64,
+    transactions_count: Option<u64>,
+    chain_id: u64,
+    cache_operator: &mut CacheOperator<redis::aio::ConnectionManager>,
+    file_store_operator: Arc<Box<dyn FileStoreOperator>>,
+    request_metadata: IndexerGrpcRequestMetadata,
+    cache_storage_format: StorageFormat,
+) -> DataFetchSubThreadResult {
+    // TODO: better logic here for when we're already caught up to head?
+    let num_threads_to_use = match transactions_count {
+        None => MAX_FETCH_THREADS_PER_REQUEST,
+        Some(transactions_count) => {
+            (transactions_count / TRANSACTIONS_PER_STORAGE_BLOCK).max(MAX_FETCH_THREADS_PER_REQUEST)
+        },
+    };
+
+    let mut threads = vec![];
+    let mut current_version = start_version;
+
+    for _ in 0..num_threads_to_use {
+        let thread = tokio::spawn({
+            // TODO: arc these instead of cloning
+            let mut cache_operator = cache_operator.clone();
+            let file_store_operator = file_store_operator.clone();
+            let request_metadata = request_metadata.clone();
+            async move {
+                get_data_in_thread(
+                    current_version,
+                    chain_id,
+                    &mut cache_operator,
+                    file_store_operator,
+                    request_metadata.clone(),
+                    cache_storage_format,
+                )
+                .await
+            }
+        });
+        threads.push(thread);
+        current_version += TRANSACTIONS_PER_STORAGE_BLOCK;
+    }
+
+    let mut transactions = vec![];
+    join_all(threads)
+        .await
+        .into_iter()
+        .for_each(|result| match result {
+            Ok(DataFetchSubThreadResult::Success(mut txns)) => {
+                transactions.append(&mut txns);
```

Comment:
We should probably check that there are no gaps in the txn's from different threads. Example of how we do it in the filestore worker: https://github.com/aptos-labs/aptos-core/blob/a808904519704c224deb67da53da86437d64ab06/ecosystem/indexer-grpc/indexer-grpc-file-store/src/processor.rs#L201

Reply:
Oh yeah that's true, there could be dupes in every thread lol